### PR TITLE
49 add ai module integration

### PR DESCRIPTION
**Pull request by inciotiq**

### CODE REVIEW

File: src/main/java/com/iotiq/application/aimodule/services/EntitiesService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.jpa.domain.AbstractPersistable;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class EntitiesService {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/aimodule/services/EntitiesService.java:19]: This service prepares a Map for the AI Module to consume

File: src/main/java/com/iotiq/application/domain/Exhibit.java [hunk 2]
```diff
@@ -1,21 +1,22 @@
...
-import org.springframework.data.jpa.domain.AbstractPersistable;
 
 import java.util.*;
 
 @Entity
 @Setter
 @Getter
-public class Exhibit extends AbstractPersistable<UUID> {
+@EqualsAndHashCode(callSuper = true)
+public class Exhibit extends NavigableObject {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/domain/Exhibit.java:15]: I extended objects that can  be navigated to from a new class called NavigableObjects and moved the location object to there.

File: src/main/java/com/iotiq/application/domain/Exhibit.java [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
     private String name;
-    @Embedded
-    private Location location;
     @OneToMany
     @JoinColumn(name = "exhibit_id")
+    @JsonIgnore
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/domain/Exhibit.java:19]: I added @JsonIgnore temporarily so the entities endpoint doesn't return an infinitely nested json response.

File: src/main/java/com/iotiq/application/domain/NavigableObject.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.data.jpa.domain.AbstractPersistable;
+
+import java.util.UUID;
+
+@Getter
+@Setter
+@Entity
+@EqualsAndHashCode(callSuper = true)
+@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
+public class NavigableObject extends AbstractPersistable<UUID> {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/domain/NavigableObject.java:19]: This class is extended by entities that can be navigated to. I used InheritanceType.TABLE_PER_CLASS so we can query each entity on its own. 

https://thorben-janssen.com/complete-guide-inheritance-strategies-jpa-hibernate/#tablePerClass

Here it says table per class is not good for polymorphic queries but I thought we would send polymorphic queries less times than we would query entities individually, so I chose this method. We can cache the polymorphic query not to use it every time the entities are requested.

File: src/main/java/com/iotiq/application/exception/TenantException.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.iotiq.application.exception;
+
+import com.iotiq.commons.exceptions.ApplicationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+import java.util.List;
+
+public class TenantException extends ApplicationException {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/exception/TenantException.java:10]: The name of the class is not great, we can change it

File: src/main/java/com/iotiq/application/util/TenantUtil.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+@Component
+@Configuration
+public class TenantUtil {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/util/TenantUtil.java:19]: I added this class in another merge request too.

File: src/main/java/com/iotiq/application/service/NavEdgeService.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.UUID;
+
+@RequiredArgsConstructor
+@Service
+public class NavEdgeService {
+
+    private final NavEdgeRepository navEdgeRepository;
+    private final NavPointRepository navPointRepository;
+
+    public UUID create(NavEdgeDto request) {
```
Comment (inciotiq) [inline src/main/java/com/iotiq/application/service/NavEdgeService.java:22]: The same edge can be created from ending point to starting point

File: src/main/java/com/iotiq/application/aimodule/AIModuleClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package com.iotiq.application.aimodule;
+
+import com.iotiq.application.aimodule.messages.NextStopRequest;
+import com.iotiq.application.aimodule.messages.NextStopResponse;
+import org.springframework.stereotype.Service;
+
+@Service
+public class AIModuleClientImpl implements AIModuleClient {
+    @Override
+    public NextStopResponse getNextStop(NextStopRequest request) {
```
[the remark this answers is not in the document]
Comment (oguzIotiq) [reply src/main/java/com/iotiq/application/aimodule/AIModuleClientImpl.java:10]: issue is opened
#96

File: src/main/java/com/iotiq/application/aimodule/AIModuleAuthService.java [hunk 9]
```diff
@@ -0,0 +1,4 @@
+package com.iotiq.application.aimodule;
+
+public interface AIModuleAuthService {
```
[the remark this answers is not in the document]
Comment (oguzIotiq) [reply src/main/java/com/iotiq/application/aimodule/AIModuleAuthService.java:3]: issue is opened
#97

File: src/main/java/com/iotiq/application/controller/FacilityController.java [hunk 10]
```diff
@@ -42,9 +42,9 @@ public ResponseEntity<FacilityDto> getOne(@PathVariable UUID id) {
 
     @PostMapping
     @PreAuthorize("hasAuthority(@MuseumManagementAuth.CREATE)")
-    public ResponseEntity<Void> create(@RequestBody @Valid FacilityCreateRequest facility) {
-        facilityService.create(facility);
-        return new ResponseEntity<>(HttpStatus.CREATED);
+    public ResponseEntity<UUID> create(@RequestBody @Valid FacilityCreateRequest facility) {
+        UUID id = facilityService.create(facility);
+        return new ResponseEntity<>(id, HttpStatus.CREATED);
```
[the remark this answers is not in the document]
Comment (oguzIotiq) [reply src/main/java/com/iotiq/application/controller/FacilityController.java:47]: #98

**Review by oguzIotiq — APPROVED** [review]

**Pull request merged**